[ATen] Support multi dim any and all reductions

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110311
* __->__ #110310

This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

peterbell10 committed:
[ATen] Support multi dim any and all reductions

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"




[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"




[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"




[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"




[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"




[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"




[ghstack-poisoned]

peterbell10 committed:
typo on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -57,11 +57,11 @@
         "amax",
         "amin",
-        "any",
+        # "any",    - onnxscript doesn't handle aten::any.dims
```
Apparently this is testing out-of-tree code, so can't be fixed in this PR
https://github.com/microsoft/onnxscript/blob/f8046e17490a3222d4ee4bca346b4449a2d30ae4/onnxscript/function_libs/torch_lib/ops/core.py#L404

peterbell10 replied:
cc @justinchuby

justinchuby replied:
The current best thing to do is to xfail it here: https://github.com/pytorch/pytorch/pull/110310/files/73dd8600a8097f98b675510024e01ea1a3d6af5d#diff-db2f78a51511bb172cbfde1b2f68272b8b33049abe2571cded27bcd0f3ae5fa4R176 and track it with an issue. 

I will discuss with the team to find a better solution forward. Thanks!

lezcano commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1541,10 +1552,46 @@ TORCH_IMPL_FUNC(any_out)
+Tensor allany_dims_default(const Tensor &self, OptionalIntArrayRef dim, bool keepdim) {
+  // Default implementation in terms of all-reduce or single dim reduce
+  if (!dim) {
+    auto out = is_all ? self.all() : self.any();
```
if constexpr, although the compiler should be able to optimise out the branch.

lezcano commented on test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:
```diff
@@ -8,17 +8,14 @@
 import parameterized
 
 import torch
+from onnx_test_common import MAX_ONNX_OPSET_VERSION, MIN_ONNX_OPSET_VERSION
```
why these changes?

peterbell10 replied:
Those constants aren't defined in the file they were being imported from. This file must not be run in CI.

lezcano replied:
mind filing an issue?

justinchuby replied:
Thanks!

lezcano commented on tools/autograd/derivatives.yaml:
```diff
@@ -295,6 +295,9 @@
 - name: any.dim(Tensor self, int dim, bool keepdim=False) -> Tensor
   output_differentiability: [False]
 
+- name: any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
+  output_differentiability: [False]
+
```
Can't you make it composite implicit, even if it's not differentiable?

peterbell10 replied:
No, because that would have higher priority than the `CPU` and `CUDA` dispatch.

lezcano commented on tools/pyi/gen_pyi.py:
```diff
@@ -171,6 +171,12 @@ def should_bind_method(python_func: PythonSignatureNativeFunctionPair) -> bool:
 
+overload_blocklist = [
+    # Overloads which are more specific than another, so not needed
+    "all.dim",
```
nit. Put the name of the other overload.

peterbell10 replied:
wdym? If we skip the array overload then mypy would complain when you pass a list of dims.

lezcano commented on torch/_refs/__init__.py:
```diff
@@ -2257,7 +2249,10 @@ def any(
+    if isinstance(dim, (list, tuple)) and len(dim) == 0:
+        result = a_.clone()
+    else:
+        result = ne(sum(a_, dim=dim, keepdim=keepdim), False)  # type: ignore[arg-type]
```
Preexisting, but we should probably be calling `torch.`  ops

lezcano commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -1541,10 +1552,46 @@ TORCH_IMPL_FUNC(any_out)
+  }
+
+  Tensor out = self;
+  for (auto d : *dim) {
```
This is fine, but couldn't we extend the tensoriterator implementation? I reckon it should be straightforward and it should be much more efficient.

peterbell10 replied:
CPU and CUDA do use tensoriterator. I included this mainly for MPS or any third-party dispatch keys.

lezcano commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -681,15 +681,29 @@
+  structured: True
+  dispatch:
+    CPU, CUDA: all_dims_out
+    MPS: all_dims_out_mps
```
does mps implement all but not any?

peterbell10 replied:
Oh sorry that's dead code. I looked briefly into adding it but the MPS code was fairly complex and I have no way to test so I decided against it. Removed in the latest commit.

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Support out overload on mps on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

justinchuby commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -681,15 +681,29 @@
   structured_delegate: all.out
   variants: function, method
 
+- func: all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```
Question: Does `int[1]` imply the length of the sequence can only be 1? I have seen the use of `int[]` too. Was this intended?

peterbell10 replied:
`int[1]` is a broadcasting list, it means if you pass a scalar it will be broadcast to a list of size 1. I'm removing the broadcasting aspect though as it's covered by the `all.dim` overload (I had hoped  we wouldn't need two).

justinchuby commented:
When `dim` is None and `keepdims` is False for `all.dims`, it is equivalent to `all.default`?

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 commented:
> When dim is None and keepdim is False for all.dims, it is equivalent to all.default? Just curious about when dim needs to take None. Would the alternative be adding keepdim=False to all.default? I assume there was a reason not to do that?

Yes, `dim=None` is an all-reduce. The reason to prefer this signature is for generic code where you accept a `dim` argument from the caller, if we can't pass `dim=None` then you need to write code like:
```python
return x.all() if dim is None else x.all(dim=dim)
```

justinchuby commented:
> > When dim is None and keepdim is False for all.dims, it is equivalent to all.default? Just curious about when dim needs to take None. Would the alternative be adding keepdim=False to all.default? I assume there was a reason not to do that?
> 
> Yes, `dim=None` is an all-reduce. The reason to prefer this signature is for generic code where you accept a `dim` argument from the caller, if we can't pass `dim=None` then you need to write code like:
> 
> ```python
> return x.all() if dim is None else x.all(dim=dim)
> ```

justinchuby commented:
Sorry wrong click

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -188,6 +188,9 @@
         reason=onnx_test_common.reason_onnx_does_not_support("Addmm")
     ),
+    xfail(
+        "all", reason=onnx_test_common.reason_onnx_script_does_not_support("aten.all.dims")
```
```suggestion
        "all", reason=onnx_test_common.reason_onnx_script_does_not_support("aten.all.dims") + "https://github.com/pytorch/pytorch/issues/XXXXX"
```

Maybe add the tracking issue as part of the reason or a `# TODO: Remove after https://github.com/pytorch/pytorch/issues/XXXXX` comment around it?

@justinchuby how do you feel changing `this to def reason_onnx_script_does_not_support(operator, dtypes, github_issue=None)` to help tracking the underlying issue. That is a way to remind users to create issues without our request

justinchuby replied:
I will update https://github.com/pytorch/pytorch/pull/110536

justinchuby commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -188,6 +188,9 @@
         "addmm", dtypes=onnx_test_common.BOOL_TYPES,
         reason=onnx_test_common.reason_onnx_does_not_support("Addmm")
     ),
+    xfail(
```
```suggestion
    xfail_torchlib_forward_compatibility(
```

Could you rebase on a commit after https://github.com/pytorch/pytorch/pull/110536 and use `xfail_torchlib_forward_compatibility`? Thanks!

justinchuby commented on test/onnx/test_fx_op_consistency.py:
```diff
@@ -201,6 +204,9 @@
         "amin", dtypes=(torch.int16, *onnx_test_common.BOOL_TYPES),
         reason=onnx_test_common.reason_dynamo_does_not_support("ReduceMin", "bool, int16")
     ),
+    xfail(
```
```suggestion
    xfail_torchlib_forward_compatibility(
```

peterbell10 commented:
@albanD do you have any idea about these segfaults in MPS? I tried disabling `CPU, CUDA` dispatch locally so I was testing with the `CompositeExplicitAutograd` version on CPU and it worked fine, so there seems to be an MPS-specific issue here.

peterbell10 commented:
@kulinseth ping

kulinseth commented:
> cc @kulinseth any idea how we can be segfaulting here? :o

It could be related to https://github.com/pytorch/pytorch/issues/95538. For >4D reduction ops are crashing.

peterbell10 commented:
Okay if it's a pre-existing issue then I'll just add a skip.

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

albanD commented:
If the error is a hard crash, you will have to add it to the skip list instead of xfail.

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot rebase

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]

peterbell10 committed:
Update on "[ATen] Support multi dim any and all reductions"


This adds a new overload to `all` and `any` with support for multiple reduction dims.
```
all.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
any.dims(Tensor self, int[1]? dim=None, bool keepdim=False) -> Tensor
```

[ghstack-poisoned]
